contribute/style-guides/e2e.md

```diff
@@ -160,3 +160,41 @@ describe('List test', () => {
   });
 });
 ```
+
+## Aria-Labels vs test-data-id
+Our selectors are set up to work with both aria-labels and data-test-id attributes. Aria-labels help assistive technologies such as screenreaders identify interactive elements of a page for our users. 
+
+A good example of a time to use an aria-label might be if you have a button with an X to close:
+```
+<button aria-label="close">X<button>
+```
+It might be clear visually that the X closes the modal, but audibly it would not be clear for example.
+```
+<button aria-label="close">Close<button>
+```
+The above example for example might read aloud to a user "Close, Close" or something similar. 
+
+However adding aria-labels to elements that are already clearly labled or not interactive can be confusing and redundant for users with assistive technologies.
```

Comment:
Good point, haven't thought about this. I'm not familiar with our e2e framework that much, but was wondering perhaps there is a way to query components through accessibility interfaces, like React Testing Library, which we use for unit tests, does? For example, in the case of button with text, we can query it with `screen.getByRole('button', {name: 'Close'})`, which doesn't require adding any test-id attribute.

Reply:
We could use something like this https://testing-library.com/docs/cypress-testing-library/intro/ but not sure if there's a way to use this with our wrappers or if we would have to use this standalone, in which case, we may have to remove the wrapper approach?

Reply:
I think it'd be great if we could make that work and it would remove the need for e2e-specific selectors.

Reply:
+1 to generally avoid e2e selectors if possible

Reply:
I guess the reasoning behind using test ids over relying on button text is that one could imagine a world in which:
1) we change the text of a button from "close" to "cancel" or something like that and then the tests break 
2) we dynamically change the text of a button with internationalization tooling and the tests break when you switch from english to spanish or something.

Using `data-test-ids` is listed as a "Best Practice" according to cypress: https://docs.cypress.io/guides/references/best-practices#Selecting-Elements

Reply:
The `data-testid` approach was discussed in a previous front end weekly meeting and we were going to replace our aria labels with them @hugohaggmark thoughts?

Reply:
oh one more reason not to do something like `screen.getByRole('button', {name: 'Close'})` is that we often have multiple buttons with the same text on the page. For example on this import dashboard page we have 2 buttons with the text Load.

I think removing our wrapper entirely so that we simply call .get directly in our tests like this:
``` js
e2e.get('[data-test-id="whatever"]') // instead of e2e.components.Thing.textarea()
```

is pretty compelling to me, as it's a bit easier to understand how everything plugs together. However I also could see a benefit of using our existing wrapper to make testing easier across different versions of grafana maybe? But I guess that felt like a bigger change that should be discussed than simply adding support for test ids

Reply:
Cypress and Testing Library have differing opinions on what to use as selectors. Generally speaking, I tend to side with using things like Testing Library where you test *as the user*. This approach works really well for "unit" tests where we're testing much smaller atoms of the interface, but like Sarah mentioned in e2e test the scope is much larger so it can be tricky to craft the right selector, which might end up being a lot more brittle.

Reply:
Yeah, all valid points, we don't really want our e2e tests to be too brittle, maybe using `testid` is the best way in this case.